nomad/client_csi_endpoint.go

```diff
@@ -109,6 +109,96 @@ func (a *ClientCSI) ControllerDetachVolume(args *cstructs.ClientCSIControllerDet
 	return fmt.Errorf("controller detach volume: %v", err)
 }
 
+func (a *ClientCSI) ControllerCreateVolume(args *cstructs.ClientCSIControllerCreateVolumeRequest, reply *cstructs.ClientCSIControllerCreateVolumeResponse) error {
+	defer metrics.MeasureSince([]string{"nomad", "client_csi_controller", "create_volume"}, time.Now())
+
+	clientIDs, err := a.clientIDsForController(args.PluginID)
+	if err != nil {
+		return fmt.Errorf("create volume: %v", err)
+	}
+
+	for _, clientID := range clientIDs {
+		args.ControllerNodeID = clientID
+		state, ok := a.srv.getNodeConn(clientID)
+		if !ok {
+			return findNodeConnAndForward(a.srv,
+				clientID, "ClientCSI.ControllerCreateVolume", args, reply)
+		}
+
+		err = NodeRpc(state.Session, "CSI.ControllerCreateVolume", args, reply)
+		if err == nil {
+			return nil
+		}
+		if a.isRetryable(err) {
+			a.logger.Debug("failed to reach controller on client",
+				"nodeID", clientID, "err", err)
+			continue
+		}
+		return fmt.Errorf("create volume: %v", err)
+	}
+	return fmt.Errorf("create volume: %v", err)
+}
+
+func (a *ClientCSI) ControllerDeleteVolume(args *cstructs.ClientCSIControllerDeleteVolumeRequest, reply *cstructs.ClientCSIControllerDeleteVolumeResponse) error {
+	defer metrics.MeasureSince([]string{"nomad", "client_csi_controller", "delete_volume"}, time.Now())
+
+	clientIDs, err := a.clientIDsForController(args.PluginID)
+	if err != nil {
+		return fmt.Errorf("controller delete volume: %v", err)
+	}
+
+	for _, clientID := range clientIDs {
+		args.ControllerNodeID = clientID
+		state, ok := a.srv.getNodeConn(clientID)
+		if !ok {
+			return findNodeConnAndForward(a.srv,
+				clientID, "ClientCSI.ControllerDeleteVolume", args, reply)
+		}
+
+		err = NodeRpc(state.Session, "CSI.ControllerDeleteVolume", args, reply)
+		if err == nil {
+			return nil
+		}
+		if a.isRetryable(err) {
```

Comment:
> block-list style 

Can you clarify what you mean by this?

Reply:
Looking at `isRetryable` I see that it deems as retryable only requests that have "unknown node" or "CSI client error (retryable)" as retryable.  I don't know how transient network or plugin failures (e.g. expired AWS creds for AWS EBS plugin) are handled - I think we'd want to continue checking other plugin nodes before giving up.

So, I'm wondering if it makes sense to have isRetryable default to treating errors as retry-able unless the error is of a known persistent failure type.

Reply:
Ah, I see what you mean. The CSI specification RPCs have a small set of common errors that are retryable but then each endpoint has its own set of possible retryable errors. Most errors in the spec are not retryable.

So for example: 
> (e.g. expired AWS creds for AWS EBS plugin)

Nomad doesn't know what that means, and any given endpoint within the specification may not actually give plugins the option to tell us that a specific kind of error is retryable (ex. https://github.com/container-storage-interface/spec/blob/master/spec.md#validatevolumecapabilities-errors only allows for a `NOT_FOUND`). So each endpoint handler in the plugin client (way down in `plugins/`) has to determine if the error is one of the retryable errors, and then we wrap it in `"CSI client error (retryable)"` error so that the servers don't have to know anything about those details.

Reply:
I see.  For my own edification - how are nomad RPC call failures handled now? If NodeRpc fails with a connection error before the request reaches the CSI plugin?

Reply:
That `NodeRpc` is the function in `nomad/client_rpc.go` that we use for things like the various non-stream Allocations RPCs. If I dig all the way down into the connection pool logic in `helpers/pool` it looks like it'll re-dial dropped connections but otherwise leaves it up to the caller to decide about things like retries. I imagine that's the right thing: the RPC library can't really know if the call is safe to retry.